hazelcast/src/main/java/com/hazelcast/crdt/CRDTMigrationTask.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.crdt;
+
+import com.hazelcast.core.Member;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+
+import java.util.Collection;
+
+import static com.hazelcast.cluster.memberselector.MemberSelectors.DATA_MEMBER_SELECTOR;
+
+/**
+ * Task responsible for migrating the CRDT states for all
+ * {@link CRDTReplicationAwareService}. This task is a runnable that is
+ * meant to be executed by an executor. The task may be interrupted in
+ * which case some CRDT states may not be replicated.
+ */
+class CRDTMigrationTask implements Runnable {
+    /** Delay in seconds for a failed migration retry */
+    private static final int MIGRATION_RETRY_DELAY_SECONDS = 1;
+    private final NodeEngine nodeEngine;
+    private final ILogger logger;
+    private final CRDTReplicationMigrationService replicationMigrationService;
+
+    CRDTMigrationTask(NodeEngine nodeEngine, CRDTReplicationMigrationService replicationMigrationService) {
+        this.nodeEngine = nodeEngine;
+        this.logger = nodeEngine.getLogger(getClass());
+        this.replicationMigrationService = replicationMigrationService;
+    }
+
+    @Override
+    public void run() {
+        try {
+            if (nodeEngine.getLocalMember().isLiteMember()) {
+                return;
+            }
+            final Collection<Member> members = nodeEngine.getClusterService().getMembers(DATA_MEMBER_SELECTOR);
+            final Member firstDataMember = members.iterator().next();
```

Comment:
can we add a comment here to describe the reasoning? If I am the first data member, there is no other member to migrate my data..

Reply:
Will do. IMO this is the biggest downside of the member list approach - the first member owns everything. The downside of the partition table based approach being that there is max 7 replicas. Maybe we can find an even better approach in the future.